include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -42,21 +44,66 @@ struct DefaultKernelName;
 template <typename KernelName = DefaultKernelName>
 class device_policy
 {
+    // Needed for the copy constructor that rebinds the kernel name
+    template <typename>
+    friend class device_policy;
+
+    template <typename T>
+    static auto
+    lock_and_forward(T&& t, std::mutex& mtx)
```

Comment:
From my point of view this pattern looks like not quite correct.
For example let's analyze the situation when we create new object ```device_policy``` by ```const device_policy&``` from another copy.
What's happen when we call ```lock_and_forward``` :
- we pass ```const device_policy&``` and mutex into this function;
- lock mutex;
- return our argument ```const device_policy&``` as result;
- unlock mutex on exit.
What we made safe by this mutex? Looks like we made safe only return of ```const device_policy&```.
I think it wasn't our goal.

Reply:
I think this pattern will work if we return simple copy of ```t```.
May be the idea was in that?

Reply:
Discussed offline that because we return ```auto``` this mean we return copy - agree, ok.

Reply:
From "Working Draft, Standard for Programming Language C++"

```
9.6.3 The return statement
.....
3 The copy-initialization of the result of the call is sequenced before the destruction of temporaries at the end of the full-expression established by the operand of the return statement, which, in turn, is sequenced before the destruction of local variables (9.6) of the block enclosing the return statement.
```

So looks like this solution with ```lock_and_forward``` static function is OK.

Reply:
@adamfidel, @rarutyun I think now we can resolve this conversation?